### PR TITLE
Fixed some issues related to the file secure storage

### PR DESCRIPTION
**Pull request by vanlooverenkoen**

### CODE REVIEW

File: lib/src/file_storage/documents_file_storage.dart [hunk 1]
```diff
@@ -1,22 +1,29 @@
...
-  String? outputPath;
-
-  FileStorage();
+class DocumentsFileStorage extends FileStorage {
+  String? _outputPath;
 
   Future<String> get documentsPath async =>
       (await getApplicationDocumentsDirectory()).path;
 
+  /// Set a custom output path, So not all files are generated at the same location.
```
Comment (NicolaVerbeeck) [inline lib/src/file_storage/documents_file_storage.dart:13]: wording and casing

File: lib/src/flutter_secure_file_storage.dart [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
-  final _fileStorage = FileStorage();
-  var _keys = <String>[];
+  late FileStorage _fileStorage;
+  final Map<String, Lock> _locks = {};
+  var _keys = <String>{};
 
   String? _outputPath;
 
-  FlutterSecureFileStorage(FlutterSecureStorage storage) {
+  // fileStorage is provided so you can use your own fileStorage implementation
```
Comment (NicolaVerbeeck) [inline lib/src/flutter_secure_file_storage.dart:21]: Extra / for docs

File: lib/src/flutter_secure_file_storage.dart [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
+import 'package:flutter_secure_file_storage/src/file_storage/documents_file_storage.dart';
 import 'package:path/path.dart';
 import 'package:flutter_secure_file_storage/src/encryption_util.dart';
-import 'package:flutter_secure_file_storage/src/file_storage.dart';
+import 'package:flutter_secure_file_storage/src/file_storage/file_storage.dart';
 import 'package:flutter_secure_file_storage/src/secure_storage.dart';
 import 'package:flutter_secure_storage/flutter_secure_storage.dart';
+import 'package:synchronized/synchronized.dart';
 
 class FlutterSecureFileStorage {
```
Comment (NicolaVerbeeck) [inline lib/src/flutter_secure_file_storage.dart:13]: Needs docs

File: lib/src/flutter_secure_file_storage.dart [hunk 4]
```diff
@@ -130,11 +143,25 @@ class FlutterSecureFileStorage {
...
   Future<void> _updateKeys() async {
     final encodedData = _keys.map((e) => base64Encode(utf8.encode(e))).toList();
-    _secureStorage.saveKeys(encodedData);
+    await _secureStorage.saveKeys(encodedData);
   }
 
   Future<void> _getKeys() async {
     final decodedData = await _secureStorage.readKeys();
-    _keys = decodedData.map((e) => utf8.decode(base64Decode(e))).toList();
+    _keys = decodedData.map((e) => utf8.decode(base64Decode(e))).toSet();
```
Comment (NicolaVerbeeck) [inline lib/src/flutter_secure_file_storage.dart:151]: Do we want the keys to be cached instead of re-read all the time?

**Review by ikbendewilliam — APPROVED** [review]

File: lib/src/flutter_secure_file_storage.dart [hunk 5]
```diff
@@ -42,49 +42,57 @@ class FlutterSecureFileStorage {
...
-        : value as Uint8List;
-    final encryptionKey = await _secureStorage.getOrGenerateKey(key);
-    final encrypted =
-        await EncryptionUtil.encrypt(encryptionKey, convertedValue);
-    if (encrypted == null) return delete(key: key);
-    await _secureStorage.saveIV(key, encrypted.iv);
-    await _fileStorage.write(_filename(key), encrypted.value);
-    _keys.add(key);
-    _updateKeys();
+    await _synchronized(key, () async {
```
Comment (NicolaVerbeeck) [inline lib/src/flutter_secure_file_storage.dart:45]: Never mind I saw the implementation below

File: lib/src/flutter_secure_file_storage.dart [hunk 6]
```diff
@@ -130,11 +143,25 @@ class FlutterSecureFileStorage {
...
+    _keys = decodedData.map((e) => utf8.decode(base64Decode(e))).toSet();
+  }
+
+  Future<T> _synchronized<T>(
+    String key,
+    FutureOr<T> Function() computation,
+  ) async {
+    final lock = _locks.putIfAbsent(key, () => Lock(reentrant: true));
+
+    final result = await lock.synchronized(() {
```
Comment (NicolaVerbeeck) [inline lib/src/flutter_secure_file_storage.dart:160]: Remove lock in finally

**Review by NicolaVerbeeck — CHANGES_REQUESTED** [review]

File: lib/src/flutter_secure_file_storage.dart (hunk 4 above)
Comment (vanlooverenkoen) [reply lib/src/flutter_secure_file_storage.dart:151]: I was thinking the same, but what if the storage_inspector changes that value? We do want to have te latest version everytime we want to use it right?

Comment (NicolaVerbeeck) [reply lib/src/flutter_secure_file_storage.dart:151]: Storage inspector has immutable keys for now ;) But yes, this probably should be fine

**Review by NicolaVerbeeck — APPROVED** [review]

**Pull request merged**